src/baseline.rs

```diff
@@ -7,22 +7,210 @@ use crate::dump::RustDocCommand;
+                },
+                CrateSource::ManifestPath { path, .. } => DependencyDetail {
+                    path: Some(
+                        // The Manifest will be saved in some other directory,
+                        // so for convenience, we're using absolute paths.
+                        path.canonicalize()
+                            .with_context(|| "failed to canonicalize Manifest path")?
+                            .to_str()
+                            .with_context(|| "Manifest path is not valid UTF-8")?
+                            .to_string(),
+                    ),
+                    features: crate_features::get_all_crate_features_from_manifest(path),
+                    ..DependencyDetail::default()
+                },
+            };
+            let mut deps = DepsSet::new();
+            deps.insert(
+                match crate_source {
+                    CrateSource::Registry { crate_ } => crate_.name().to_string(),
+                    CrateSource::ManifestPath { name, path: _ } => name.to_string(),
+                },
+                Dependency::Detailed(project_with_features),
+            );
+            deps
+        },
+        ..Default::default()
+    })
+}
+
+fn save_placeholder_rustdoc_manifest(
+    placeholder_build_dir: &Path,
+    placeholder_manifest: cargo_toml::Manifest<()>,
+) -> anyhow::Result<PathBuf> {
+    std::fs::create_dir_all(placeholder_build_dir).with_context(|| "failed to create build dir")?;
+    let placeholder_manifest_path = placeholder_build_dir.join("Cargo.toml");
+
+    // Possibly fixes https://github.com/libp2p/rust-libp2p/pull/2647#issuecomment-1280221217
+    let _: std::io::Result<()> = std::fs::remove_file(placeholder_build_dir.join("Cargo.lock"));
+
+    std::fs::write(
+        &placeholder_manifest_path,
+        toml::to_string(&placeholder_manifest)?,
+    )
+    .with_context(|| "failed to write placeholder Manifest")?;
+    std::fs::write(placeholder_build_dir.join("lib.rs"), "")
+        .with_context(|| "failed to create empty lib.rs")?;
+    Ok(placeholder_manifest_path)
+}
+
+fn generate_rustdoc(
+    config: &mut GlobalConfig,
+    rustdoc: &RustDocCommand,
+    target_root: PathBuf,
+    crate_source: CrateSource,
+) -> anyhow::Result<PathBuf> {
+    let (name, version) = match crate_source {
+        CrateSource::Registry { crate_ } => (crate_.name().to_string(), crate_.version()),
+        CrateSource::ManifestPath { .. } => unimplemented!(),
+    };
+    let (build_dir, cache_dir) = match crate_source {
+        CrateSource::Registry { .. } => {
+            let crate_identifier = format!("registry-{}-{}", slugify(&name), slugify(version));
+            let cache_dir = target_root.join("cache").join(&crate_identifier);
+
+            // We assume that the generated rustdoc is untouched.
+            // Users should run cargo-clean if they experience any anomalies.
+            if cache_dir.exists() {
```

Comment:
We need to check for existence of the JSON file, not the directory.

Reply:
From what I remember, I've changed the code a bit so that the cached rustdoc is in the format `cache/registry-xxx-yyy/rustdoc.json` instead of `cache/registry-xxx-yyy.json`, with the reasoning that the code would be a bit easier to understand after the changes I've made (because of the `let (x, y) = z`), and I won't deny it -- this change is mostly stylistic. I think the code you've pointed out after my changes is correct -- can you confirm?

Reply:
Oh, I see that @obi1kenobi has also commented this code. I'll look into it in detail.

Reply:
I'll revert to how it was before.

Reply:
I don't think the directory-per-file layout is desirable, and I since we already have something that we think works and that we want to test in the real world, I think we probably don't want to change the behavior here and now.

Reply:
Your PR description said "no functionality was changed" which is why I commented :)

I think a flatter file hierarchy is preferable but I am obviously biased 😁

Whatever you go with, test it locally please that it still works as we don't have any automated tests for this!

Reply:
42f0fb99e99f332da156f6dac7d10f663dd0d0bb reverts it. Thanks for the help.

Reply:
> Your PR description said "no functionality was changed" which is why I commented :)

Right, that totally was my mistake. I'm still learning :)

Reply:
> we don't have any automated tests for this

By a happy accident, we actually do — and they caught a minor bug:
https://github.com/obi1kenobi/cargo-semver-checks/actions/runs/4079374830/jobs/7030684022

That test runs effectively the same command over the same crate in two different ways (since they were broken due to separate reasons), and it happens to run back to back in the same runner. So we can see that the caching worked (yay! 🙌) but we can also see that when the cache gets hit, the `(baseline)` printout is just inexplicably gone (confusing 🤨).
![image](https://user-images.githubusercontent.com/2348618/216481260-738cf659-f676-4d90-a0fd-61ecb4cac877.png)

Let's add a user-facing message there that indicates the cache hit. Suggestion: `Parsing libp2p-core v0.37.0 (baseline, cached)`

Reply:
@thomaseizinger you might find the subject crate in the test familiar 😁

Reply:
> Let's add a user-facing message there that indicates the cache hit. Suggestion: `Parsing libp2p-core v0.37.0 (baseline, cached)`

Done in 54e8b67b3fb46fc69fbb24207ecf102ba38f59c6.